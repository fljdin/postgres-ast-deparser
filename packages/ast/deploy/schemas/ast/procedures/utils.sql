-- Deploy schemas/ast/procedures/utils to pg

-- requires: schemas/ast/schema

BEGIN;


COMMIT;
