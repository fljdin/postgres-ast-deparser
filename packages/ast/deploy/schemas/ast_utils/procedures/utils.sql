-- Deploy schemas/ast_utils/procedures/utils to pg

-- requires: schemas/ast_utils/schema

BEGIN;

CREATE FUNCTION ast_utils.interval (
  n int
) returns text[] as $$
	select (CASE 
WHEN ( n = 2 ) THEN ARRAY[ 'month' ]
WHEN ( n = 4 ) THEN ARRAY[ 'year' ]
WHEN ( n = 6 ) THEN ARRAY[ 'year', 'month' ]
WHEN ( n = 8 ) THEN ARRAY[ 'day' ]
WHEN ( n = 1024 ) THEN ARRAY[ 'hour' ]
WHEN ( n = 1032 ) THEN ARRAY[ 'day', 'hour' ]
WHEN ( n = 2048 ) THEN ARRAY[ 'minute' ]
WHEN ( n = 3072 ) THEN ARRAY[ 'hour', 'minute' ]
WHEN ( n = 3080 ) THEN ARRAY[ 'day', 'minute' ]
WHEN ( n = 4096 ) THEN ARRAY[ 'second' ]
WHEN ( n = 6144 ) THEN ARRAY[ 'minute', 'second' ]
WHEN ( n = 7168 ) THEN ARRAY[ 'hour', 'second' ]
WHEN ( n = 7176 ) THEN ARRAY[ 'day', 'second' ]
WHEN ( n = 32767 ) THEN ARRAY[]::text[]
END);
$$  
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.reserved (
  str text
) returns boolean as $$
	select exists( select 1 from pg_get_keywords() where catcode = 'R' AND word=str  );
$$  
LANGUAGE 'sql' SECURITY DEFINER;

CREATE FUNCTION ast_utils.objtypes ()
returns text[] as $$
	select ARRAY[ 'ACCESS METHOD', 'AGGREGATE', NULL, NULL, NULL, 'CAST', 'COLUMN', 'COLLATION', 'CONVERSION', 'DATABASE', NULL, NULL, 'DOMAIN', 'CONSTRAINT', NULL, 'EXTENSION', 'FOREIGN DATA WRAPPER', 'SERVER', 'FOREIGN TABLE', 'FUNCTION', 'INDEX', 'LANGUAGE', 'LARGE OBJECT', 'MATERIALIZED VIEW', 'OPERATOR CLASS', 'OPERATOR', 'OPERATOR FAMILY', 'POLICY', NULL, NULL, 'ROLE', 'RULE', 'SCHEMA', 'SEQUENCE', NULL, 'STATISTICS', 'CONSTRAINT', 'TABLE', 'TABLESPACE', 'TRANSFORM', 'TRIGGER', 'TEXT SEARCH CONFIGURATION', 'TEXT SEARCH DICTIONARY', 'TEXT SEARCH PARSER', 'TEXT SEARCH TEMPLATE', 'TYPE', NULL, 'VIEW' ]::text[];
$$  
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.constrainttype_idxs (typ text)
returns int as $$
  select (CASE
WHEN (typ = 'CONSTR_NULL') THEN 0
WHEN (typ = 'CONSTR_NOTNULL') THEN 1
WHEN (typ = 'CONSTR_DEFAULT') THEN 2
WHEN (typ = 'CONSTR_IDENTITY') THEN 3
WHEN (typ = 'CONSTR_CHECK') THEN 4
WHEN (typ = 'CONSTR_PRIMARY') THEN 5
WHEN (typ = 'CONSTR_UNIQUE') THEN 6
WHEN (typ = 'CONSTR_EXCLUSION') THEN 7
WHEN (typ = 'CONSTR_FOREIGN') THEN 8
WHEN (typ = 'CONSTR_ATTR_DEFERRABLE') THEN 9
WHEN (typ = 'CONSTR_ATTR_NOT_DEFERRABLE') THEN 10
WHEN (typ = 'CONSTR_ATTR_DEFERRED') THEN 11
WHEN (typ = 'CONSTR_ATTR_IMMEDIATE') THEN 12
END);
$$
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.constrainttypes (contype int)
returns text as $$
  select (CASE
WHEN (contype =  0 ) THEN 'NULL'
WHEN (contype =  1 ) THEN 'NOT NULL'
WHEN (contype =  2 ) THEN 'DEFAULT'
WHEN (contype =  4 ) THEN 'CHECK'
WHEN (contype =  5 ) THEN 'PRIMARY KEY'
WHEN (contype =  6 ) THEN 'UNIQUE'
WHEN (contype =  7 ) THEN 'EXCLUDE'
WHEN (contype =  8 ) THEN 'REFERENCES'
END);
$$
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.objtypes_idxs (typ text)
returns int as $$
	select (CASE
WHEN (typ = 'OBJECT_ACCESS_METHOD') THEN 0
WHEN (typ = 'OBJECT_AGGREGATE') THEN 1
WHEN (typ = 'OBJECT_AMOP') THEN 2
WHEN (typ = 'OBJECT_AMPROC') THEN 3
WHEN (typ = 'OBJECT_ATTRIBUTE') THEN 4
WHEN (typ = 'OBJECT_CAST') THEN 5
WHEN (typ = 'OBJECT_COLUMN') THEN 6
WHEN (typ = 'OBJECT_COLLATION') THEN 7
WHEN (typ = 'OBJECT_CONVERSION') THEN 8
WHEN (typ = 'OBJECT_DATABASE') THEN 9
WHEN (typ = 'OBJECT_DEFAULT') THEN 10
WHEN (typ = 'OBJECT_DEFACL') THEN 11
WHEN (typ = 'OBJECT_DOMAIN') THEN 12
WHEN (typ = 'OBJECT_DOMCONSTRAINT') THEN 13
WHEN (typ = 'OBJECT_EVENT_TRIGGER') THEN 14
WHEN (typ = 'OBJECT_EXTENSION') THEN 15
WHEN (typ = 'OBJECT_FDW') THEN 16
WHEN (typ = 'OBJECT_FOREIGN_SERVER') THEN 17
WHEN (typ = 'OBJECT_FOREIGN_TABLE') THEN 18
WHEN (typ = 'OBJECT_FUNCTION') THEN 19
WHEN (typ = 'OBJECT_INDEX') THEN 20
WHEN (typ = 'OBJECT_LANGUAGE') THEN 21
WHEN (typ = 'OBJECT_LARGEOBJECT') THEN 22
WHEN (typ = 'OBJECT_MATVIEW') THEN 23
WHEN (typ = 'OBJECT_OPCLASS') THEN 24
WHEN (typ = 'OBJECT_OPERATOR') THEN 25
WHEN (typ = 'OBJECT_OPFAMILY') THEN 26
WHEN (typ = 'OBJECT_POLICY') THEN 27
WHEN (typ = 'OBJECT_PUBLICATION') THEN 28
WHEN (typ = 'OBJECT_PUBLICATION_REL') THEN 29
WHEN (typ = 'OBJECT_ROLE') THEN 30
WHEN (typ = 'OBJECT_RULE') THEN 31
WHEN (typ = 'OBJECT_SCHEMA') THEN 32
WHEN (typ = 'OBJECT_SEQUENCE') THEN 33
WHEN (typ = 'OBJECT_SUBSCRIPTION') THEN 34
WHEN (typ = 'OBJECT_STATISTIC_EXT') THEN 35
WHEN (typ = 'OBJECT_TABCONSTRAINT') THEN 36
WHEN (typ = 'OBJECT_TABLE') THEN 37
WHEN (typ = 'OBJECT_TABLESPACE') THEN 38
WHEN (typ = 'OBJECT_TRANSFORM') THEN 39
WHEN (typ = 'OBJECT_TRIGGER') THEN 40
WHEN (typ = 'OBJECT_TSCONFIGURATION') THEN 41
WHEN (typ = 'OBJECT_TSDICTIONARY') THEN 42
WHEN (typ = 'OBJECT_TSPARSER') THEN 43
WHEN (typ = 'OBJECT_TSTEMPLATE') THEN 44
WHEN (typ = 'OBJECT_TYPE') THEN 45
WHEN (typ = 'OBJECT_USER_MAPPING') THEN 46
WHEN (typ = 'OBJECT_VIEW') THEN 47
END);
$$  
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.getgrantobject (node jsonb)
returns text as $$
DECLARE
  objtype int;
  targtype int;
BEGIN 

  objtype = (node->'objtype')::int;
  IF (node->'targtype') IS NOT NULL THEN
    targtype = (node->'targtype')::int;
  END IF;

  IF (objtype = 0) THEN
    RETURN 'COLUMN';
  ELSIF (objtype = 1) THEN
    IF (targtype = 1) THEN 
      RETURN 'ALL TABLES IN SCHEMA';
    ELSIF (targtype = 2) THEN 
      RETURN 'TABLES';
    END IF;
    -- TODO could be a view
    RETURN 'TABLE';
  ELSIF (objtype = 2) THEN
    RETURN 'SEQUENCE';
  ELSIF (objtype = 3) THEN
    RETURN 'DATABASE';
  ELSIF (objtype = 4) THEN
    RETURN 'DOMAIN';
  ELSIF (objtype = 5) THEN
    RETURN 'FOREIGN DATA WRAPPER';
  ELSIF (objtype = 6) THEN
    RETURN 'FOREIGN SERVER';
  ELSIF (objtype = 7) THEN
    IF (targtype = 1) THEN 
      RETURN 'ALL FUNCTIONS IN SCHEMA';
    ELSIF (targtype = 2) THEN 
      RETURN 'FUNCTIONS';
    END IF;
    RETURN 'FUNCTION';
  ELSIF (objtype = 8) THEN
    RETURN 'LANGUAGE';
  ELSIF (objtype = 9) THEN
    RETURN 'LARGE OBJECT';
  ELSIF (objtype = 10) THEN
    RETURN 'SCHEMA';
  ELSIF (objtype = 11) THEN
    RETURN 'TABLESPACE';
  ELSIF (objtype = 12) THEN
    RETURN 'TYPE';
  END IF;

  RAISE EXCEPTION 'BAD_EXPRESSION %', 'GrantObjectType';

END;
$$  
LANGUAGE 'plpgsql' IMMUTABLE;

COMMIT;
