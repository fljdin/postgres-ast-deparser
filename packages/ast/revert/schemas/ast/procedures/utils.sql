-- Revert schemas/ast/procedures/utils from pg

BEGIN;

DROP FUNCTION ast.utils;

COMMIT;
