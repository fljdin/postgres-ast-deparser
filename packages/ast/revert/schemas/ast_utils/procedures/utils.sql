-- Revert schemas/ast_utils/procedures/utils from pg

BEGIN;

DROP FUNCTION ast_utils.utils;

COMMIT;
