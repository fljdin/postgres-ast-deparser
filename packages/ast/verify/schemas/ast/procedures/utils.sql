-- Verify schemas/ast/procedures/utils  on pg

BEGIN;

SELECT verify_function ('ast.utils');

ROLLBACK;
