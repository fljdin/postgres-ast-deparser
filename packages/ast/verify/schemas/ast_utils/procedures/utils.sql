-- Verify schemas/ast_utils/procedures/utils  on pg

BEGIN;

SELECT verify_function ('ast_utils.utils');

ROLLBACK;
