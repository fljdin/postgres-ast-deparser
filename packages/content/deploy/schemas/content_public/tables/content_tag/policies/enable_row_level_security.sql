-- Deploy schemas/content_public/tables/content_tag/policies/enable_row_level_security to pg

-- requires: schemas/content_public/schema
-- requires: schemas/content_public/tables/content_tag/table

BEGIN;

ALTER TABLE content_public.content_tag
    ENABLE ROW LEVEL SECURITY;

COMMIT;
