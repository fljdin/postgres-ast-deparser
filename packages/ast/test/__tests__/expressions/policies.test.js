import { getConnections } from '../../utils';
import { policies } from './__fixtures__/policies';
import policyStmt from './__fixtures__/policy';

let db, teardown;
const objs = {
  tables: {}
};

beforeAll(async () => {
  ({ db, teardown } = await getConnections());
  await db.begin();
  await db.savepoint();
});
afterAll(async () => {
  try {
    //try catch here allows us to see the sql parsing issues!
    await db.rollback();
    await db.commit();
    await teardown();
  } catch (e) {}
});

it('deparse', async () => {
  const [{ deparse: result }] = await db.any(`
  select deparser.deparse( '${JSON.stringify(policies[0])}'::jsonb );
  `);
  expect(result).toMatchSnapshot();
});

it('policy deparse', async () => {
  const result = await db.any(`
SELECT deparser.deparse(ast_helpers.create_policy(
  v_policy_name := 'mypolicy',
  v_schema_name := 'schemanamed',
  v_table_name := 'mytable',
  v_roles := '{authenticated}'::text[],
  v_qual := ast.bool_expr(1, to_jsonb(ARRAY[
    ast.a_expr(v_kind := 0,
      v_lexpr := ast.column_ref(
        v_fields := to_jsonb(ARRAY[ ast.string('responder_id') ])
      ),
      v_name := to_jsonb(ARRAY[ast.string('=')]),
      v_rexpr := ast.func_call(
        v_funcname := to_jsonb(ARRAY[ ast.string('dbe'), ast.string('get_uid') ]),
        v_args := to_jsonb(ARRAY[ ast.string('c'), ast.string('b') ])
      )  
    ),
    ast.a_expr(v_kind := 0,
      v_lexpr := ast.column_ref(
        v_fields := to_jsonb(ARRAY[ ast.string('requester_id') ])
      ),
      v_name := to_jsonb(ARRAY[ast.string('=')]),
      v_rexpr := ast.func_call(
        v_funcname := to_jsonb(ARRAY[ ast.string('dbe'), ast.string('get_other_uid') ]),
        v_args := to_jsonb(ARRAY[ ast.string('c'), ast.string('b') ])
      )  
    )
  ])),
  v_cmd_name := 'INSERT',
  v_with_check := NULL,
  v_permissive := true
))`);
  expect(result).toMatchSnapshot();
});

const getPolicyResult = async (name, vars) => {
  const result = await db.any(
    `
SELECT deparser.deparse(
  ast_helpers.create_policy_template(
  rls_schema := 'rls_schema',
  role_fn := 'role_fn',
  groups_fn := 'group_fn',
  policy_template_name := $1::text,
  policy_template_vars := $2::jsonb
  ))`,
    [name, JSON.stringify(vars)]
  );
  return result;
};

it('own_records', async () => {
  const result = await getPolicyResult('own_records', {
    role_key: 'role_key'
  });
  expect(result).toMatchSnapshot();
});

it('owned_records', async () => {
  const result = await getPolicyResult('owned_records', {
    role_key: 'role_key'
  });
  expect(result).toMatchSnapshot();
});

it('multi_owners', async () => {
  const result = await getPolicyResult('multi_owners', {
    role_keys: ['requester_id', 'responder_id', 'verifier_id']
  });
  expect(result).toMatchSnapshot();
});

it('permission_name', async () => {
  const result = await getPolicyResult('permission_name', {
    permission_role_key: 'permission_role_key',
    permission_schema: 'permission_schema',
    permission_table: 'permission_table',
    permission_name_key: 'permission_name_key',
    this_value: 'this_value'
  });
  expect(result).toMatchSnapshot();
});

it('owned_object_records', async () => {
  const result = await getPolicyResult('owned_object_records', {
    owned_table_key: 'owned_table_key',
    owned_schema: 'owned_schema',
    owned_table: 'owned_table',
    owned_table_ref_key: 'owned_table_ref_key',
    this_object_key: 'this_object_key'
  });
  expect(result).toMatchSnapshot();
});

it('child_of_owned_object_records', async () => {
  const result = await getPolicyResult('child_of_owned_object_records', {
    owned_table_key: 'owned_table_key',
    object_schema: 'object_schema',
    object_table: 'object_table',
    owned_schema: 'owned_schema',
    owned_table: 'owned_table',
    owned_table_ref_key: 'owned_table_ref_key',
    object_table_owned_key: 'object_table_owned_key',
    object_table_ref_key: 'object_table_ref_key',
    this_object_key: 'this_object_key'
  });
  expect(result).toMatchSnapshot();
});

it('child_of_owned_object_records_with_ownership', async () => {
  const result = await getPolicyResult(
    'child_of_owned_object_records_with_ownership',
    {
      owned_table_key: 'owned_table_key',
      object_schema: 'object_schema',
      object_table: 'object_table',
      owned_schema: 'owned_schema',
      owned_table: 'owned_table',
      owned_table_ref_key: 'owned_table_ref_key',
      object_table_owned_key: 'object_table_owned_key',
      object_table_ref_key: 'object_table_ref_key',
      this_object_key: 'this_object_key',

      // only one extra?
      this_owned_key: 'this_owned_key'
    }
  );
  expect(result).toMatchSnapshot();
});

it('open', async () => {
  const result = await getPolicyResult('open', {});
  expect(result).toMatchSnapshot();
});

it('policyStmt', async () => {
  const result = await db.any(
    `
SELECT deparser.deparse(
    $1::jsonb
  )`,
    [JSON.stringify(policyStmt)]
  );
  expect(result).toMatchSnapshot();
});

it('drop_policy', async () => {
  const result = await db.any(
    `
SELECT deparser.deparse(
  ast_helpers.drop_policy(
    v_policy_name := 'v_policy_name',
    v_schema_name := 'v_schema_name',
    v_table_name := 'v_table_name'
  )
  )`
  );
  expect(result).toMatchSnapshot();
});
