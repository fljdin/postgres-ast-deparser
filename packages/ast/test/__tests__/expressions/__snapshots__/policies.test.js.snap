// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`child_of_owned_object_records 1`] = `
Object {
  "insert": "CREATE POLICY v_policy_name ON v_schema_name.v_table_name FOR INSERT TO authenticated WITH CHECK ( (SELECT p.owned_table_key = ANY( rls_schema.group_fn() ) FROM object_schema.object_table AS c INNER JOIN owned_schema.owned_table AS p ON p.owned_table_ref_key = c.object_table_owned_key WHERE c.object_table_ref_key = this_object_key) );",
  "policy": "(SELECT p.owned_table_key = ANY( rls_schema.group_fn() ) FROM object_schema.object_table AS c INNER JOIN owned_schema.owned_table AS p ON p.owned_table_ref_key = c.object_table_owned_key WHERE c.object_table_ref_key = this_object_key)",
  "update": "CREATE POLICY v_policy_name ON v_schema_name.v_table_name FOR UPDATE TO authenticated USING ( (SELECT p.owned_table_key = ANY( rls_schema.group_fn() ) FROM object_schema.object_table AS c INNER JOIN owned_schema.owned_table AS p ON p.owned_table_ref_key = c.object_table_owned_key WHERE c.object_table_ref_key = this_object_key) );",
}
`;

exports[`child_of_owned_object_records_group_array 1`] = `
Object {
  "insert": "CREATE POLICY v_policy_name ON v_schema_name.v_table_name FOR INSERT TO authenticated WITH CHECK ( (SELECT rls_schema.role_fn() = ANY( g.owned_table_key ) FROM messages_schema.messages_table AS m INNER JOIN groups_schema.groups_table AS g ON g.id = m.group_id WHERE m.id = id) );",
  "policy": "(SELECT rls_schema.role_fn() = ANY( g.owned_table_key ) FROM messages_schema.messages_table AS m INNER JOIN groups_schema.groups_table AS g ON g.id = m.group_id WHERE m.id = id)",
  "update": "CREATE POLICY v_policy_name ON v_schema_name.v_table_name FOR UPDATE TO authenticated USING ( (SELECT rls_schema.role_fn() = ANY( g.owned_table_key ) FROM messages_schema.messages_table AS m INNER JOIN groups_schema.groups_table AS g ON g.id = m.group_id WHERE m.id = id) );",
}
`;

exports[`child_of_owned_object_records_with_ownership 1`] = `
Object {
  "insert": "CREATE POLICY v_policy_name ON v_schema_name.v_table_name FOR INSERT TO authenticated WITH CHECK ( (SELECT p.owned_table_key = ANY( rls_schema.group_fn() ) FROM object_schema.object_table AS c INNER JOIN owned_schema.owned_table AS p ON p.owned_table_ref_key = c.object_table_owned_key AND p.owned_table_ref_key = this_owned_key WHERE c.object_table_ref_key = this_object_key) );",
  "policy": "(SELECT p.owned_table_key = ANY( rls_schema.group_fn() ) FROM object_schema.object_table AS c INNER JOIN owned_schema.owned_table AS p ON p.owned_table_ref_key = c.object_table_owned_key AND p.owned_table_ref_key = this_owned_key WHERE c.object_table_ref_key = this_object_key)",
  "update": "CREATE POLICY v_policy_name ON v_schema_name.v_table_name FOR UPDATE TO authenticated USING ( (SELECT p.owned_table_key = ANY( rls_schema.group_fn() ) FROM object_schema.object_table AS c INNER JOIN owned_schema.owned_table AS p ON p.owned_table_ref_key = c.object_table_owned_key AND p.owned_table_ref_key = this_owned_key WHERE c.object_table_ref_key = this_object_key) );",
}
`;

exports[`deparse 1`] = `"CREATE POLICY authenticated_can_delete_on_user_secrets ON dashboard_private.user_secrets FOR DELETE TO authenticated USING ( user_id = dashboard_public.get_current_user_id() );"`;

exports[`drop_policy 1`] = `
Array [
  Object {
    "deparse": "DROP POLICY v_policy_name ON v_schema_name.v_table_name;",
  },
]
`;

exports[`multi_owners 1`] = `
Object {
  "insert": "CREATE POLICY v_policy_name ON v_schema_name.v_table_name FOR INSERT TO authenticated WITH CHECK ( requester_id = rls_schema.role_fn() OR responder_id = rls_schema.role_fn() OR verifier_id = rls_schema.role_fn() );",
  "policy": "requester_id = rls_schema.role_fn() OR responder_id = rls_schema.role_fn() OR verifier_id = rls_schema.role_fn()",
  "update": "CREATE POLICY v_policy_name ON v_schema_name.v_table_name FOR UPDATE TO authenticated USING ( requester_id = rls_schema.role_fn() OR responder_id = rls_schema.role_fn() OR verifier_id = rls_schema.role_fn() );",
}
`;

exports[`open 1`] = `
Object {
  "insert": "CREATE POLICY v_policy_name ON v_schema_name.v_table_name FOR INSERT TO authenticated WITH CHECK ( TRUE );",
  "policy": "TRUE",
  "update": "CREATE POLICY v_policy_name ON v_schema_name.v_table_name FOR UPDATE TO authenticated USING ( TRUE );",
}
`;

exports[`own_records 1`] = `
Object {
  "insert": "CREATE POLICY v_policy_name ON v_schema_name.v_table_name FOR INSERT TO authenticated WITH CHECK ( role_key = rls_schema.role_fn() );",
  "policy": "role_key = rls_schema.role_fn()",
  "update": "CREATE POLICY v_policy_name ON v_schema_name.v_table_name FOR UPDATE TO authenticated USING ( role_key = rls_schema.role_fn() );",
}
`;

exports[`own_records defaults 1`] = `
Object {
  "insert": "CREATE POLICY v_policy_name ON v_schema_name.v_table_name FOR INSERT TO authenticated WITH CHECK ( role_key = jwt_public.current_user_id() );",
  "policy": "role_key = jwt_public.current_user_id()",
  "update": "CREATE POLICY v_policy_name ON v_schema_name.v_table_name FOR UPDATE TO authenticated USING ( role_key = jwt_public.current_user_id() );",
}
`;

exports[`owned_object_records 1`] = `
Object {
  "insert": "CREATE POLICY v_policy_name ON v_schema_name.v_table_name FOR INSERT TO authenticated WITH CHECK ( (SELECT p.owned_table_key = ANY( rls_schema.group_fn() ) FROM owned_schema.owned_table AS p WHERE p.owned_table_ref_key = this_object_key) );",
  "policy": "(SELECT p.owned_table_key = ANY( rls_schema.group_fn() ) FROM owned_schema.owned_table AS p WHERE p.owned_table_ref_key = this_object_key)",
  "update": "CREATE POLICY v_policy_name ON v_schema_name.v_table_name FOR UPDATE TO authenticated USING ( (SELECT p.owned_table_key = ANY( rls_schema.group_fn() ) FROM owned_schema.owned_table AS p WHERE p.owned_table_ref_key = this_object_key) );",
}
`;

exports[`owned_object_records_group_array 1`] = `
Object {
  "insert": "CREATE POLICY v_policy_name ON v_schema_name.v_table_name FOR INSERT TO authenticated WITH CHECK ( (SELECT rls_schema.role_fn() = ANY( p.owned_table_key ) FROM owned_schema.owned_table AS p WHERE p.owned_table_ref_key = this_object_key) );",
  "policy": "(SELECT rls_schema.role_fn() = ANY( p.owned_table_key ) FROM owned_schema.owned_table AS p WHERE p.owned_table_ref_key = this_object_key)",
  "update": "CREATE POLICY v_policy_name ON v_schema_name.v_table_name FOR UPDATE TO authenticated USING ( (SELECT rls_schema.role_fn() = ANY( p.owned_table_key ) FROM owned_schema.owned_table AS p WHERE p.owned_table_ref_key = this_object_key) );",
}
`;

exports[`owned_records 1`] = `
Object {
  "insert": "CREATE POLICY v_policy_name ON v_schema_name.v_table_name FOR INSERT TO authenticated WITH CHECK ( role_key = rls_schema.role_fn() OR role_key = ANY( rls_schema.group_fn() ) );",
  "policy": "role_key = rls_schema.role_fn() OR role_key = ANY( rls_schema.group_fn() )",
  "update": "CREATE POLICY v_policy_name ON v_schema_name.v_table_name FOR UPDATE TO authenticated USING ( role_key = rls_schema.role_fn() OR role_key = ANY( rls_schema.group_fn() ) );",
}
`;

exports[`owned_records defaults 1`] = `
Object {
  "insert": "CREATE POLICY v_policy_name ON v_schema_name.v_table_name FOR INSERT TO authenticated WITH CHECK ( role_key = jwt_public.current_user_id() OR role_key = ANY( jwt_public.current_group_ids() ) );",
  "policy": "role_key = jwt_public.current_user_id() OR role_key = ANY( jwt_public.current_group_ids() )",
  "update": "CREATE POLICY v_policy_name ON v_schema_name.v_table_name FOR UPDATE TO authenticated USING ( role_key = jwt_public.current_user_id() OR role_key = ANY( jwt_public.current_group_ids() ) );",
}
`;

exports[`permission_name 1`] = `
Object {
  "insert": "CREATE POLICY v_policy_name ON v_schema_name.v_table_name FOR INSERT TO authenticated WITH CHECK ( (SELECT p.permission_role_key = ANY( rls_schema.group_fn() ) FROM permission_schema.permission_table AS p WHERE p.permission_name_key = 'this_value') );",
  "policy": "(SELECT p.permission_role_key = ANY( rls_schema.group_fn() ) FROM permission_schema.permission_table AS p WHERE p.permission_name_key = 'this_value')",
  "update": "CREATE POLICY v_policy_name ON v_schema_name.v_table_name FOR UPDATE TO authenticated USING ( (SELECT p.permission_role_key = ANY( rls_schema.group_fn() ) FROM permission_schema.permission_table AS p WHERE p.permission_name_key = 'this_value') );",
}
`;

exports[`policy deparse 1`] = `
Array [
  Object {
    "deparse": "CREATE POLICY mypolicy ON schemanamed.mytable FOR INSERT TO authenticated USING ( responder_id = dbe.get_uid(c, b) OR requester_id = dbe.get_other_uid(c, b) );",
  },
]
`;

exports[`policy w permission_name 1`] = `"CREATE POLICY v_policy_name ON v_schema_name.v_table_name FOR INSERT TO authenticated WITH CHECK ( (SELECT p.permission_role_key = ANY( rls_schema.group_fn() ) FROM permission_schema.permission_table AS p WHERE p.permission_name_key = 'this_value') );"`;

exports[`policyStmt 1`] = `
Array [
  Object {
    "deparse": "CREATE POLICY delete_the_thing ON users.\\"user\\" FOR DELETE TO PUBLIC USING ( (SELECT p.owned_table_key = ANY( current_role_fn_schema.current_group_ids_function() ) FROM object_schema.object_table AS c INNER JOIN owned_schema.owned_table AS p ON p.owned_table_ref_key = c.object_table_owned_key AND p.owned_table_ref_key = this_owned_key WHERE c.object_table_ref_key = this_object_key) );",
  },
]
`;
